pkg/controllers/machine/link/suite_test.go

```diff
@@ -80,9 +97,30 @@ var _ = BeforeSuite(func() {
 	ctx = settings.ToContext(ctx, test.Settings())
 	env = coretest.NewEnvironment(scheme.Scheme, coretest.WithCRDs(apis.CRDs...))
 	unavailableOfferingsCache = awscache.NewUnavailableOfferings()
+	launchTemplateCache = cache.New(awscache.DefaultTTL, awscache.DefaultCleanupInterval)
+	ssmCache = cache.New(awscache.DefaultTTL, awscache.DefaultCleanupInterval)
+	ec2Cache = cache.New(awscache.DefaultTTL, awscache.DefaultCleanupInterval)
+	kubernetesVersionCache = cache.New(awscache.DefaultTTL, awscache.DefaultCleanupInterval)
+	instanceTypeCache = cache.New(awscache.DefaultTTL, awscache.DefaultCleanupInterval)
 	ec2API = &fake.EC2API{}
+	ssmAPI = &fake.SSMAPI{}
 	subnetProvider = subnet.NewProvider(ec2API)
+	securityGroupProvider = securitygroup.NewProvider(ec2API)
 	pricingProvider = pricing.NewProvider(ctx, &fake.PricingAPI{}, ec2API, "", make(chan struct{}))
+	amiProvider = amifamily.NewProvider(env.Client, env.KubernetesInterface, ssmAPI, ec2API, ssmCache, ec2Cache, kubernetesVersionCache)
+	amiResolver = amifamily.New(env.Client, amiProvider)
+	launchTemplateProvider = launchtemplate.NewProvider(
+		ctx,
+		launchTemplateCache,
+		ec2API,
+		amiResolver,
+		securityGroupProvider,
+		ptr.String("ca-bundle"),
+		make(chan struct{}),
+		net.ParseIP("10.0.100.10"),
+		"https://test-cluster",
+	)
+	instanceProvider = instance.NewProvider(ctx, "", ec2API, unavailableOfferingsCache, instanceTypeProvider, subnetProvider, launchTemplateProvider)
 	instanceTypesProvider = instancetype.NewProvider("", cache.New(awscache.DefaultTTL, awscache.DefaultCleanupInterval), ec2API, subnetProvider, unavailableOfferingsCache, pricingProvider)
 	cloudProvider = cloudprovider.New(awscontext.Context{
```

Comment:
I think we only need to instantiate the things in these tests that the test has dependency on. So in this case, I'm not sure that we need to take dependency on all of these other providers

Reply:
Unfortunately there is a dependency between the providers. So the extra providers that are instantiated are need for the providers that the tests has a dependency on.